lib/perf.js

```diff
@@ -0,0 +1,196 @@
+      const sep = baseURL.at(-1) !== '/' && o.url?.[0] !== '/' ? '/' : ''
+      o.url = /^https?:/.test(o.url) ? o.url : `${baseURL}${sep}${o.url}`
+    }
+    return o
+  }
+
+  async _run(opts) {
+    this.autocannon
+
+    let { fn, args } = opts
+    if (args) fn = fn.bind(null, ...args)
+
+    if (opts.warmup) await this._run({ ...opts, fn, args: undefined, ...opts.warmup, warmup: undefined })
+
+    const { getHistograms, encodeHist } = this._histUtil
+
+    const histograms = getHistograms(opts.histograms)
+    const { latencies, requests, throughput } = histograms
+
+    const statusCodeStats = {}
+
+    let stop = false
+    let count = 0
+    let errors = 0
+    let nextTrack
+    let totalRequests = 0
+    let totalCompletedRequests = 0
+
+    const runners = new Array(opts.connections)
+
+    const startTime = process.hrtime.bigint()
+    const endTime = startTime + BigInt((typeof opts.duration === 'string' ? this._timestring(opts.duration) : opts.duration) * 1e9)
+
+    for (let r = 0; r < runners.length; r++) {
+      runners[r] = run()
+    }
+    await Promise.all(runners)
+
+    const result = {
+      latencies: encodeHist(latencies),
+      requests: encodeHist(requests),
+      throughput: encodeHist(throughput),
+      totalCompletedRequests,
+      totalRequests,
+      totalBytes: 0,
+      samples: Math.floor(Number(process.hrtime.bigint() - startTime) / 1e9),
+      errors,
+      timeouts: 0,
+      mismatches: 0,
+      non2xx: Object.keys(statusCodeStats).reduce((l, c) => l + (c[0] === '2' ? 0 : statusCodeStats[c]), 0),
+      statusCodeStats,
+      resets: 0,
+      duration: Number(process.hrtime.bigint() - startTime) / 1e9,
+      start: new Date(Number(startTime)),
+      finish: new Date(),
+      '1xx': 0,
+      '2xx': statusCodeStats['200']?.count || 0,
+      '3xx': 0,
+      '4xx': 0,
+      '5xx': statusCodeStats['500']?.count || 0,
+    }
+
+    return this._aggregateResult(result, opts, histograms)
+
+    async function run() {
+      while (!stop) {
+        const now = process.hrtime.bigint()
+        if (!nextTrack) nextTrack = now + BigInt(1e9)
+        if (now >= nextTrack) {
+          nextTrack = now + BigInt(1e9)
+          requests.recordValue(count)
+          count = 0
+        }
+
+        if (now >= endTime) {
+          stop = true
+          break
+        }
+
+        totalRequests++
+        try {
+          const s = process.hrtime.bigint()
+
+          const ret = fn()
+          if (ret?.then) await ret;
+
+          const d = process.hrtime.bigint() - s
+          latencies.recordValue(Number(d) / 1000000)
+
+          count++
+          totalCompletedRequests++
+          (statusCodeStats['200'] ??= { count: 0 }).count++
+        } catch {
+          errors++
+          (statusCodeStats['500'] ??= { count: 0 }).count++
+        }
+      }
+    }
+  }
+
+  async _report(result, options = {}) {
+    let { requests, latency, throughput, title = `${this._reports++}` } = result
```

Comment:
smt like this would be nice to provide a title for one report
```suggestion
    let { requests, latency, throughput, title = options.title || `${this._reports++}` } = result
```